### PR TITLE
Enhance types and composability

### PR DESCRIPTION
**Pull request by pcbowers**
TLDR: Adds better types, improves extensibility, and enhances composability.

- Found a way to infer types straight from an interface (i.e. Basically inferred generics!). This makes type hover infinitely better, and also decreases the number of types that need to be passed around by a substantial amount
- Included a function called `assign` that is essentially a direct improvement on `Object.assign`. According to the [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#copying_accessors), `Object.assign` does not copy accessors, and instead takes the value of a getter and setter and moves it directly into an object. Furthermore, `Object.assign` only assigns own properties. Since class instances don't own methods, `Object.assign` cannot be used to combine two instances of a class. This new `assign` function allows class instances and objects to be combined along with class methods while simultaneously improving the overall type inference.
- While it most likely won't be used a whole lot, I wanted to support other people extending the prototype. I added a static `extend` method to the BaseRegExp so that new methods/properties/getters can be added easily. Type inference can also still work using a `declare module` statement. See the tests for an example.

### CODE REVIEW

**Comment by pcbowers** [comment]
Closes #9
